### DISCUSSION TITLE
Register View in Svelte

### DISCUSSION THREAD

**Discussion by TQuantab** [toeverything/blocksuite#5568] [Q&A] [answered]
Hi there :)
im currently trying to make Blocksuite work in my small svelte Project.

I already managed to get the "default" editor to work, and im now trying to add a custom block. So far so good, the custom block is loading and can be added via a button. 

My Problem is now, that it wont get rendered. Im always getting a "Cannot find view for my-block." error. 
I registered the Schema  together with the default affine Schema. Do i also have to register the view somewhere (its in the same file as the schema)?

The view/element works, if i try to use it it normal html.

Thanks for your help =)

**Comment by doodlewind** [comment]
Hi, thanks for your trying! We haven't exposed the custom block helper API (prototype recently implemented in #5518) in our docs yet, which should allow you to define `embed-something` block easily. Could you share more about the code snippet or API used for customizing? Thanks.

**Reply by TQuantab** [reply]
Hi!
thank you for your fast answer =D

I've defined a Block in its own Folder with a Schema, a view and a Service. (Each in its own file).

**The Schema:**

```js
export const MyBlockSchema = defineBlockSchema({
	flavour: 'my-block',
	props: (internal) => ({
		text: internal.Text(),
		count: 0
	}),
	metadata: {
		version: 1,
		role: 'content',
		parent: ['affine:note', 'affine:paragraph']
	}
});

export type MyBlockModel = SchemaToModel<typeof MyBlockSchema>;`
```

**The Service:**
```js
 export class MyBlockService extends BlockService<MyBlockModel> {
	    override mounted() {
		    super.mounted();
		    this.bindHotkey(
			    {
				    'Alt-1': this._addMyBlock
			    },
			    { global: true }
		    );
	    }
    
	    private _addMyBlock = () => {
		    this.page.addBlock('my-block', {});
	    };
    }
```

**And the View:**
```js
@customElement('my-block')
export class MyBlockView extends BlockElement<MyBlockModel> {
	private _yen = '0¥';

	private _onClick = () => {
		this.page.updateBlock(this.model, {
			count: this.model.count + 1
		});
	};
	override connectedCallback() {
		super.connectedCallback();

		this.model.propsUpdated.on(() => {
			this._yen = `${this.model.count * 100}¥`;
		});
	}
	override render() {
		return html`
			<div>
				<h3>My Block</h3>
				<p>Price: ${this._yen}</p>
				<button @click=${this._onClick}>Add</button>
			</div>
		`;
	}
}

declare global {
	interface HTMLElementTagNameMap {
		'my-block': MyBlockView;
	}
}
```

**And then i added the schema, when creating my editor:**
```js
const schema = new Schema().register(AffineSchemas).register([MyBlockSchema]);
const workspace = new Workspace({ id: 'foo', schema });
```

**And used it to add the Block to the Page**
```js
page.load(() => {
    const pageBlockId = page.addBlock('affine:page', {text: new Text("my page")});
    noteId = page.addBlock('affine:note', {}, pageBlockId);
    const paraId = page.addBlock('affine:paragraph', {text: new Text("my Paragraph")}, noteId);
    page.addBlock('my-block',{}, paraId)
});
```

I've also created a Blockspec:
```js
const myBlock = {
        schema: MyBlockSchema,
        service: MyBlockService,
        view: {
            component: literal`my-block`,
            widgets: {
            },
        },
    };
```
    
But i couldn't find where to use it.
    
My Problem is now, that while i can register the view and add the Block to the Page, its not being rendered, because it says it cant find the view.

**The Errors:**
```js
Uncaught TypeError: Cannot read properties of undefined (reading 'view')
    at MyBlockView.connectedCallback (block-element.ts:153:27)
    at MyBlockView.connectedCallback (myblock.ts:16:9)
    at append_hydration (dom.js:223:12)
    at insert_hydration (dom.js:251:3)
    at insert_hydration_dev (dev.js:65:2)
    at mount (+page.svelte:66:1)
    at targetCmp.$$.fragment.m (svelte-hooks.js?v=39c640fc:291:24)
    at mount_component (Component.js:44:23)
    at Object.mount [as m] (root.svelte:44:82)
    at mount (+layout.svelte:2:24)
```

**And**
```js
Cannot find view for my-block.
```

**Reply by Flrande** [reply]
> I've also created a Blockspec:
> 
> ```ts
> const myBlock = {
>         schema: MyBlockSchema,
>         service: MyBlockService,
>         view: {
>             component: literal`my-block`,
>             widgets: {
>             },
>         },
>     };
> ```
> 
> But i couldn't find where to use it.

You need to load the spec into the root. Currently, you can refer to the implementation of `editor-container` for the rendering method of blocks. We plan to split it into two editors (corresponding to doc mode and edgeless mode).

https://github.com/toeverything/blocksuite/blob/129ec7e560f2d827e97afc49c47b2af401e13db7/packages/presets/src/components/editor-container.ts#L217

You can find the currently corresponding presets for two modes in the blocks package.

https://github.com/toeverything/blocksuite/blob/129ec7e560f2d827e97afc49c47b2af401e13db7/packages/blocks/src/_presets/index.ts#L40

**Reply by TQuantab** [reply]
Thanks for the reply :)

If i understand this correctly, this is where i would have to add my spec:

```js
export const PagePreset: BlockSpec[] = [
  DocPageSpec,
  {
    schema: SurfaceBlockSchema,
    view: {
      component: literal`affine-surface`,
    },
    service: SurfaceService,
  },
  {
    schema: ListBlockSchema,
    view: {
      component: literal`affine-list`,
    },
  },
  {
    schema: NoteBlockSchema,
    service: NoteService,
    view: {
      component: literal`affine-note`,
    },
  },
  {
    schema: DatabaseBlockSchema,
    service: DatabaseService,
    view: {
      component: literal`affine-database`,
    },
  },
  {
    schema: DataViewBlockSchema,
    view: {
      component: literal`affine-data-view`,
    },
  },
  {
    schema: DividerBlockSchema,
    view: {
      component: literal`affine-divider`,
    },
  },
  {
    schema: CodeBlockSchema,
    view: {
      component: literal`affine-code`,
    },
  },
  {
    schema: ImageBlockSchema,
    service: ImageService,
    view: {
      component: literal`affine-image`,
      widgets: {
        imageToolbar: literal`affine-image-toolbar-widget`,
      },
    },
  },
  {
    schema: ParagraphBlockSchema,
    view: {
      component: literal`affine-paragraph`,
    },
  },
  {
    schema: BookmarkBlockSchema,
    view: {
      component: literal`affine-bookmark`,
    },
  },
  {
    schema: AttachmentBlockSchema,
    view: {
      component: literal`affine-attachment`,
    },
    service: AttachmentService,
  },
  {
    schema: SurfaceRefBlockSchema,
    view: {
      component: literal`affine-surface-ref`,
      widgets: {
        surfaceToolbar: literal`affine-surface-ref-toolbar`,
      },
    },
  },
  EmbedGithubBlockSpec,
];
```

How would i do that with my current implementation of the editor:
```js
    import '@blocksuite/editor/themes/affine.css';

    import { Workspace, Schema, Text } from '@blocksuite/store';
    import { AffineSchemas } from '@blocksuite/blocks/models';
    import {EditorContainer} from '@blocksuite/editor';
    import {onMount} from "svelte";
    import {MyBlockView, MyBlockSchema, MyBlockService} from "$lib/editor/blocks/myblock";


    const schema = new Schema().register(AffineSchemas).register([MyBlockSchema]);
    const workspace = new Workspace({ id: 'foo', schema });

    const page = workspace.createPage({ id: 'page0' });
    let noteId;


    page.load(() => {
        const pageBlockId = page.addBlock('affine:page', {text: new Text("my page")});
        noteId = page.addBlock('affine:note', {}, pageBlockId);
        const paraId = page.addBlock('affine:paragraph', {text: new Text("my Paragraph")}, noteId);
        page.addBlock('my-block',{}, paraId)
    });


    page.slots.blockUpdated.on((e)=>{
        if(e.type == "add") {
            console.log("block updated");
        }

    })

    const editor = new EditorContainer();

    onMount(()=>{
        const ew = document.getElementById("editor-wrapper");
        editor.page = page;
        editor.mode = "page";
        ew.appendChild(editor);
    })

    function switchMode() {
        if (editor instanceof EditorContainer) {
            editor.mode = editor.mode === 'page' ? 'edgeless' : 'page';
        } else {
            editor.mode = editor.mode === 'page' ? 'edgeless' : 'page';
        }
    }

    function addBlock(){
        page.addBlock('my-block',{text: new Text("Paragraph")},noteId)
    }
```

Im currently not using my own Editorcontainer, im just loading in the one from @blocksuite/editor.

Would i have to create my own implementation? Should i load the source into my project, instead of using the "normal" npm release, in order to directly modify the files more easily?
(sorry if the question is dumb, im not the most experienced, when it comes to github)

**Reply by Flrande** [reply]
You can override `pagePreset` and `edgelessPreset` in `EditorContainer`:

https://github.com/toeverything/blocksuite/blob/75625432a27f3571036173a24bf0ed6d24ffdb80/packages/presets/src/components/editor-container.ts#L55-L59

Here is how we use `EditorContainer` in playground:

https://github.com/toeverything/blocksuite/blob/75625432a27f3571036173a24bf0ed6d24ffdb80/packages/playground/apps/default/utils.ts#L72-L90

**Reply by TQuantab** [reply]
This worked flawlessly =D

Last question:
is the edgeless Mode part of the "default" editor? When i change page.mode to edgeless, i only get errors.

```js
Uncaught (in promise) TypeError: Cannot destructure property 'flavour' of 'model' as it is undefined.
    at BlockSuiteRoot2.renderModel (lit-root.ts:126:13)
    at BlockElement.renderModel (block-element.ts:141:22)
    at EdgelessPageBlockComponent2.render (edgeless-page-block.ts:864:24)
    at EdgelessPageBlockComponent2.update (lit-element.ts:158:24)
    at EdgelessPageBlockComponent2.update (edgeless-page-block.ts:578:11)
    at EdgelessPageBlockComponent2.performUpdate (reactive-element.ts:1445:14)
    at EdgelessPageBlockComponent2.scheduleUpdate (reactive-element.ts:1342:25)
    at EdgelessPageBlockComponent2.__enqueueUpdate (reactive-element.ts:1314:25)
```

**Reply by Flrande** [reply]
Changing the mode of the editor-container from page to edgeless simply involves changing the loaded preset from PagePreset to EdgelessPreset, so you need to add your custom block spec to both presets. However, I'm not sure what effect the slash menu has caused. Will there be any errors if it is enabled?

https://github.com/toeverything/blocksuite/blob/be90383a3083c490242968ceb892eb8a5ab6b588/packages/presets/src/components/editor-container.ts#L217

**Reply by TQuantab** [reply]
Already solved the slash menu, it was caused by this:
```js
this.bindHotkey(
			{
				'Alt-1': this._addMyBlock
			},
			{ global: true }
		);
```

Your doc says, i should use bindHotKey. But this does not seem to be defined.

**Reply by TQuantab** [reply]
Just tested adding my spec to the edgelesspreset and removing my customizations from everything.

This error:
```js
lit-root.ts:126 Uncaught (in promise) TypeError: Cannot destructure property 'flavour' of 'model' as it is undefined.
    at BlockSuiteRoot2.renderModel (lit-root.ts:126:13)
    at BlockElement.renderModel (block-element.ts:141:22)
    at EdgelessPageBlockComponent2.render (edgeless-page-block.ts:864:24)
    at EdgelessPageBlockComponent2.update (lit-element.ts:158:24)
    at EdgelessPageBlockComponent2.update (edgeless-page-block.ts:578:11)
    at EdgelessPageBlockComponent2.performUpdate (reactive-element.ts:1445:14)
    at EdgelessPageBlockComponent2.scheduleUpdate (reactive-element.ts:1342:25)
    at EdgelessPageBlockComponent2.__enqueueUpdate (reactive-element.ts:1314:25)
```

On a clean implementation, without any changes from my side. Edglessmode seems to be missing something

**Reply by TQuantab** [reply]
im also getting this error for edgeless:
```js
editor-container.ts:289 Uncaught (in promise) TypeError: Cannot destructure property 'viewport' of 'edgelessPage.surface' as it is null.
    at EditorContainer2._saveViewportLocalRecord (editor-container.ts:289:15)
    at EditorContainer2.updated (editor-container.ts:192:14)
    at EditorContainer2._$didUpdate (reactive-element.ts:1494:10)
    at EditorContainer2.performUpdate (reactive-element.ts:1459:12)
    at EditorContainer2.scheduleUpdate (reactive-element.ts:1342:25)
    at EditorContainer2.__enqueueUpdate (reactive-element.ts:1314:25)
    ```

**Reply by Flrande** [reply]
This might be due to the absence of a surface block (editor needs it in edgeless mode). Did you add a surface block when initializing the data?

Like that:
https://github.com/toeverything/blocksuite/blob/be90383a3083c490242968ceb892eb8a5ab6b588/packages/playground/apps/default/main.ts#L139

**Reply by TQuantab** [reply]
This did the trick!! =D

Last (two) Questions, and then i should be fully able to work the rest out by myself:
1) How can i modify the "/" command view, to add my custom blocks to it (and maybe remove others)?
2) I found out, that there is a page.addSiblingBlocks. Is there also some function, to just add one block below another? The table block always blocks the editor, because it does not automatically add another paragraph below it, meaning you cant continue writing. I would like to automatically add another block below it, when its added via "page.slots.blockUpdated"

**Reply by Flrande** [reply]
> How can i modify the "/" command view, to add my custom blocks to it (and maybe remove others)?

Currently, you can achieve this by overriding the options of the slash-menu. You can refer to the following PR and tests. However, it's important to note that this API is likely to undergo restructuring in the future, but the basic customization approach will remain the same, which is to achieve it by overriding the existing widgets in the preset.

https://github.com/toeverything/blocksuite/pull/3407

https://github.com/toeverything/blocksuite/blob/86224046343e4c182eca18ba0179caf371a5bc4e/tests/slash-menu.spec.ts#L637-L743

>  found out, that there is a page.addSiblingBlocks. Is there also some function, to just add one block below another? The table block always blocks the editor, because it does not automatically add another paragraph below it, meaning you cant continue writing. I would like to automatically add another block below it, when its added via "page.slots.blockUpdated"

You can use the `parentIndex` in `addBlock` to specify the insertion position of the block. Most of these functions can be found on `Page`.

https://github.com/toeverything/blocksuite/blob/be90383a3083c490242968ceb892eb8a5ab6b588/packages/store/src/workspace/page.ts#L325-L330

**Reply by TQuantab** [reply]
Thanks! Is there an easy way to get the index of a block? I couldnt find it on the block, when using getBlockById.

**Reply by Flrande** [reply]
We usually use `parent.children.indexOf(model)`.

You can refer to the code in this file:
https://github.com/toeverything/blocksuite/blob/be90383a3083c490242968ceb892eb8a5ab6b588/packages/blocks/src/_common/components/rich-text/rich-text-operations.ts#L103

**Reply by TQuantab** [reply]
This works =)

im just getting a minor error (even though it works as intended)

```js
+page.svelte:22 Error: Length exceeded!
    at create3 (error.js:12:28)
    at lengthExceeded (AbstractType.js:686:36)
    at typeListInsertGenerics (AbstractType.js:699:11)
    at YArray.js:145:9
    at transact (Transaction.js:425:14)
    at _YArray.insert (YArray.js:144:7)
    at page.ts:362:19
    at transact (Transaction.js:425:14)
    at _Doc.transact (Doc.js:167:12)
    at Page.transact (space.ts:115:21)
```

When using this code snippet:
```js
    page.slots.blockUpdated.on((e)=>{
        //add empty line below "blocking" elements
        if(e.type == "add" && e.flavour !== "affine:paragraph"){
            const addedBlock = page.getBlockById(e.id);
            const parent = page.getParent(addedBlock);
            const index = parent.children.indexOf(addedBlock);

            page.addBlock('affine:paragraph',{}, noteId, index + 1);
        }
    });
 ```

**Reply by Flrande** [reply]
Maybe you should use `page.addBlock('affine:paragraph', {}, parent, index + 1);`

**Reply by TQuantab** [reply]
Same error as before :/ It works though!

Is there a central way to access the page object? From another file? One of the examples you gave me used "window.$blocksuite".
Which doesnt seem to work anymore.

Im currently retrieving the page, by adding it to the window Object on creation. And whenever i want to add a Block inside its actions, i do as follow:
```js
export function addMyBlock() {
	const myWindow = window as unknown as customWindowObject;
	const page = myWindow.editor.page;

	//find the page._blocks that has a value.flavour of "affine:note"
	const noteBlock = page.getBlockByFlavour('affine:note');
	const noteId = noteBlock[0].id;

	page.addBlock(page.addBlock('my-block', {}, noteId));
}
```

which works, but has the weird effect, that sometimes, the note cannot be found

**Reply by Flrande** [reply]
We bind it in windows just for tests, and it is not a good practice in fact.

https://github.com/toeverything/blocksuite/blob/f5c712f9a264058f7ee11c0d4e7cf9272a2f81f3/packages/playground/apps/starter/utils.ts#L71-L80

Where do you want to access page object? If you just want to add a paragraph after some special blocks, you can add the logic in page-service. Service is a global singleton for specific blocks in the editor, with some lifecycle functions available for use. You can extend the default page service and add it in preset (like the widget).

https://github.com/toeverything/blocksuite/blob/be90383a3083c490242968ceb892eb8a5ab6b588/packages/blocks/src/page-block/page-service.ts#L24-L53

**Reply by TQuantab** [reply]
Im accessing the page object inside my "myblock-actions.ts". I though about separating the actions concerning a block, in its own file.

So when i use my custom "Slash Menu" item, i just have to call the "addMyBlock" function to add a Block. So i need a way to access the Page object inside the file. Thats why i used the windows object.

**Reply by Flrande** [reply]
Why not wrap them as a function and pass `pageElement` or `model` as an argument into them?

https://github.com/toeverything/blocksuite/blob/dd29665fb233222412b7b4d42adc7df551c70e4b/packages/blocks/src/_common/widgets/slash-menu/config.ts#L62-L614

**Reply by TQuantab** [reply]
Didnt even think about that :0
This works flawlessly, thank you so much =D

Im still getting this two errors though, no idea why. Everything works as aspected:

```js
+page.svelte:21 Error: Invalid schema for affine:paragraph: Hub/Content must have parent.
    at Schema.validate (schema.ts:81:13)
    at Page.addBlock (page.ts:333:17)
    at +page.svelte:34:18
    at slot.ts:106:9
    at Array.forEach (<anonymous>)
    at _Slot.emit (slot.ts:104:21)
    at Page._handleYBlockAdd (page.ts:737:29)
    at page.ts:762:16
    at Map.forEach (<anonymous>)
    at Page._handleYEvent (page.ts:760:18)

+page.svelte:22 Error: Length exceeded!
    at create3 (error.js:12:28)
    at lengthExceeded (AbstractType.js:686:36)
    at typeListInsertGenerics (AbstractType.js:699:11)
    at YArray.js:145:9
    at transact (Transaction.js:425:14)
    at _YArray.insert (YArray.js:144:7)
    at page.ts:362:19
    at transact (Transaction.js:425:14)
    at _Doc.transact (Doc.js:167:12)
    at Page.transact (space.ts:115:21)
```

"+page.svelte:21-22" would be:
```js
20: const pageBlockId = page.addBlock('affine:page', {text: new Text("my page")});
21: noteId = page.addBlock('affine:note', {}, pageBlockId);
22: page.addBlock('affine:surface', {}, pageBlockId);
23: page.addBlock('affine:paragraph', {text: new Text("my Paragraph")}, noteId);
24: page.addBlock('my-block',{}, noteId)
```

"+page.svelte:34" would be:
```js
29: if(e.type == "add" && e.flavour !== "affine:paragraph"){
30:      const addedBlock = page.getBlockById(e.id);
31:       const parent = page.getParent(addedBlock); 
32:      const index = parent.children.indexOf(addedBlock);
33:
34:      page.addBlock('affine:paragraph',{}, noteId, index + 1);
}
```

according to the Documentation, this should be correct. 
"page.addBlock('affine:paragraph',{}, noteId, index + 1);" => doesnt it have a parent, being the note block?

**Reply by TQuantab** [reply]
Theres also a Problem with "renderModelChildren", that is implemented in the BlockElement. Even though my BlockView extens Blockelement, i cant access this.renderModelChildren. Any idea, what im doing wrong?

```js
export class MyBlockView extends BlockElement<MyBlockModel> {
	private _yen = '0¥';

	private _onClick = () => {
		this.page.updateBlock(this.model, {
			count: this.model.count + 1
		});
	};
	override connectedCallback() {
		super.connectedCallback();

		this.model.propsUpdated.on(() => {
			this._yen = `${this.model.count * 100}¥`;
		});
	}
	override render() {
		return html`
			<div class="">
				<div>
					<h3>My Block</h3>
					<p>Price: ${this._yen}</p>
					<button @click=${this._onClick}>Add</button>
				</div>
				<div>${this.renderModelChildren(this.model)}</div>
			</div>
		`;
	}
}
```

**Reply by TQuantab** [reply, marked as answer]
Manged to find the error: this.renderModelChildren seems to be the new way, and is not yet pushed to release? My NPM pacakge still uses ${this.content}

**Comment by MentalGear** [comment]
I'm also interested in getting this to work with svelte. Could you maybe give an outline of your setup, or is there now an official guide for this?

**Reply by TQuantab** [reply]
Im planning on rewriting the view/Ui in svelte and using blocksuite headless, as soon as the documentation is a bit more streamlined.

What I'm doing currently is using the slot/event system to communicate between my svelte setup and the "default" Lit interface.

So for the entire Editor part, I'm using lit, and lit components (as laid out in the docs). Whenever something happens, that needs to work with my svelte app (like assigning a user, to a todo), I fire an event that then triggers a svelte function (ex. show user select popUp).

**Reply by MentalGear** [reply]
Thx for the info! Yeah slot/events/attr was also my idea for connecting with svelte.
Did you encounter any problems with this setup, or is it just not flexible enough that you consider rolling your own editor components?

If you get to implementing a headless version with svelte, let me know if you're going to make it a package. :)

On a somewhat related note, (as I'm also new to Yjs), do you know of good resources on how to handle auth in Yjs?

**Reply by TQuantab** [reply]
We did not really encounter any Problem. We just want to switch to Svelte, so everything is unified, and we can use our component libraries for the Editor. At the same time, writing it from 0, gives us more control overall.

Y.js authentication could be done via websocket. We personally dont handle it ourselfs and use liveblocks "https://liveblocks.io/" for the Collaboration. They have there own auth handlers, in which you can dock you custom authentication logic.

If we write something on our own, i will send you a msg =)

**Reply by MentalGear** [reply]
Sure, you can also write it here for others to get the info :)
